Display rejected games in a table

blackshadowshade opened the pull request:
Fixes #1747. Fixes #1758. Fixes #1760. Fixes #1761.

Depends on #1743.

This has been minimally tested for the ability to display rejected games and dismiss rejected games from the list.

It is intentional that rejecting a game does not automatically dismiss it as well.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/928/ (if it passes)

blackshadowshade commented:
I'll update the pull for you once #1743 has been accepted, so that you get a clean diff.

cgolubi1 commented:
Drat.  This has merge conflicts.

cgolubi1 commented:
Hmm, looks like i can test this even though i can't merge it.  That's weird.

cgolubi1 commented:
Is there some reason the rejected games table doesn't include Max Wins even though the other tables do?  It feels distracting, different for the sake of being different.  Why not default to consistency?

![many_tables](https://cloud.githubusercontent.com/assets/2898207/10865655/0ef9b844-7fe9-11e5-8fff-fe22df8cdc7d.png)

cgolubi1 commented on src/api/DummyApiResponder.php:
```diff
@@ -704,6 +723,48 @@ protected function get_interface_response_loadCompletedGames() {
+            'opponentColorArray' => array(),
+        );
+
+        // game 5
```
Nit: you list this as game 505, so the comment should say game 505 too.  :>)

cgolubi1 commented:
Okay, so the UI draws a horizontal line above Active games... but this pull put rejected games below that horizontal line (and keeps New and Completed games above it).  That's obviously not right.  Can you move Rejected games above the line to go with New and Completed games?
![wheres_that_hr](https://cloud.githubusercontent.com/assets/2898207/10865697/053b6db4-7feb-11e5-86a4-d14bc9dee29d.png)

dwvanstone commented:
In my opinion, if I'm the one who created a game it seems odd to have the option to "Reject" it in the New Games table. I think the better verb in this case is "Cancel" and it could immediately disappear. (It seems like an extra step for me to Reject it and then to have to Dismiss it.) Rejected Tables would then only be ones that somebody else declined the invitation.

cgolubi1 commented:
> In my opinion, if I'm the one who created a game it seems odd to have the option to "Reject" it in the New Games table. I think the better verb in this case is "Cancel" and it could immediately disappear. (It seems like an extra step for me to Reject it and then to have to Dismiss it.) Rejected Tables would then only be ones that somebody else declined the invitation.

I agree, but i also think that this is an edge case, not a prod-blocker, and the logical change to implement it could add more complexity and need more testing.  So my preference would be to make this a separate issue, and not block this pull on it.

cgolubi1 commented:
For anyone watching this: i can't merge this, and thus can't cleanly load it on dev, because of the merge conflict, but i was able to test it using vagrant, and i stood up a temporary public instance so other people could test it.  If you want to test this fix and i haven't e-mailed you, ping me, and i'll let you know where the temporary instance is.

dwvanstone commented:
Just to be clear, what is the edge case you're talking about? When I create a game I will see a table that gives me the option to "Reject" it. That verb might be confusing.

dwvanstone commented:
I think I know - the edge case is that if you do press it you will have to also dismiss it. (The other part I was compalining about.) I agree that is not the standard case.

cgolubi1 commented:
Yes, exactly, that's the edge case i meant.  For those playing the home game, i created #1754 for the "Reject should say Cancel" issue.

cgolubi1 commented:
To summarize the requests so far:
* The merge conflict needs to be fixed
* Nit: comment typo in DummyApiResponder.php
* The Rejected Games table is missing the "Max Wins" column --- bring the code in from the "New Games" table to generate that column for consistency
* The horizontal line the UI draws above Active Games needs to be below Rejected Games rather than above it
* If you cancel a game you yourself created, it will show up in Rejected Games for you.  It should simply disappear from your view.  (We propose making a separate issue for this which is not a prod-blocker)

irilyth commented:
Some comments!

> The Rejected Games table is missing the "Max Wins" column --- bring the code in from the "New Games" table to generate that column for consistency

An alternative idea: Why not make the columns the same in all the tables? Except for the final column, which is either Action, Completed, Rejected, or Inactivity, I think it'd be nice if all the others were the same across all of them. (So the New and Rejected tables should both have "Score // W/L/T (Max)" rather than just "Max wins"; among other things, it makes it easy to tell at a glance what that column is, since "0/0/0 (3)" is instantly recognizable in a way that just "3" isn't.

> The horizontal line the UI draws above Active Games needs to be below Rejected Games rather than above it

Actually, I think it'd make more sense to have Rejected, Completed, the line, New, Active. That way, everything above the line is stuff that's in your court, but that no one but you cares about; and everything below the line is something that other people care about, with things that are in your court first (the second half of the Active table, with games that aren't in your court but that someone else cares about, is at the bottom, which seems right to me).

cgolubi1 commented:
> Actually, I think it'd make more sense to have Rejected, Completed, the line, New, Active

That sounds fine to me.

cgolubi1 commented:
Updating the summary, folding in irilyth's comments:

* The merge conflict needs to be fixed
* Nit: comment typo in DummyApiResponder.php
* The Rejected Games table is missing the "Max Wins" column --- all tables should have that column, and consider using the identical "Score (Max Wins)" syntax in all tables
* The horizontal line the UI draws above Active Games needs to be below Rejected Games rather than above it --- possibly address this by making the table ordering "Rejected", "Completed", line, "New", "Active"
* If you cancel a game you yourself created, it will show up in Rejected Games for you. It should simply disappear from your view.  This has been split into #1761, and is not a blocker, but would be neat to address if it turned out to be trivial.

blackshadowshade commented:
Okay, I've addressed all the points listed above. For the point about the horizontal line, I've removed it entirely for the moment. If we decide that we really need it, we'll need a new issue for that. However, I think that it adds visual complexity without actually having much benefit, especially since there are potentially five or six different tables present on the overview page at any time.

I've also rolled in fixes for #1761, #1760, and #1758, and I'll update the description at the top to add those fixes there.

http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/930/ (if it passes)

cgolubi1 commented:
In #1761, irilyth said: I think ideally the "Successfully rejected game text at the top" should instead say "Successfully canceled Game N", with a link, so that if you clicked cancel by mistake, you could click the link, get the game page, and click something like "create a new game against with the same parameters"...

Obviously, this fix doesn't do that (and can't until #1731 is addressed).  Is your expectation that he should open a new issue with that request?  Or should we modify this pull to note that it only partially addresses #1761.

blackshadowshade commented:
Ah, I think I forgot to uncomment the expect(3) when I was getting the QUnit test to run. It does work.

As for the second part of #1761, I didn't notice that because it was in brackets.  :) I think that we should open a new issue with an explicitly different name (and possibly split it so that some of the content is added to #1731), so that it's easier to work out what is still to be done.

cgolubi1 commented:
Yes, i do still want the horizontal line (or some equivalent visual differentiator) --- opened #1763

cgolubi1 commented:
> The merge conflict needs to be fixed

Fixed

> Nit: comment typo in DummyApiResponder.php

Fixed

> The Rejected Games table is missing the "Max Wins" column --- all tables should have that column, and consider using the identical "Score (Max Wins)" syntax in all tables

The column now appears in all tables.  It is still "Max Wins" in the Rejected and New tables --- irilyth, if you still think it should be "Score (Max Wins)" in all tables, maybe open a separate issue for that?  IMO it's not a blocker, and i'd like to merge this pull if i can.

> The horizontal line the UI draws above Active Games needs to be below Rejected Games rather than above it --- possibly address this by making the table ordering "Rejected", "Completed", line, "New", "Active"

Minimally done, opened #1763 for the rest

> If you cancel a game you yourself created, it will show up in Rejected Games for you. It should simply disappear from your view. This has been split into #1761, and is not a blocker, but would be neat to address if it turned out to be trivial.

Again, this was minimally done, and shadowshade requested that irilyth open a new issue with the parts of #1761 which were not yet done

cgolubi1 commented:
Jenkins for expect: http://jenkins.buttonweavers.com:8080/job/buttonmen-blackshadowshade/935/

cgolubi1 commented:
I'm not planning to wait for jenkins to finish --- i'll go ahead and merge, and start a pull to the dev site.  If it fails, we'll have another pull to fix the test.
